rust/connlib/l4-udp-dns-server/lib.rs

```diff
@@ -0,0 +1,262 @@
+                    .send_to(response.as_slice(), to)
+                    .await
+                    .context("Failed to send UDP response")?;
+
+                Ok(())
+            }
+            .boxed(),
+        );
+
+        Ok(())
+    }
+
+    pub fn poll(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<Query>> {
+        loop {
+            if let Poll::Ready(Some(result)) = self.sending_udp_v4_responses.poll_next_unpin(cx) {
+                result
+                    .context("Failed to send UDPv4 DNS response")
+                    .map_err(anyhow_to_io)?;
+
+                continue;
+            }
+
+            if let Poll::Ready(Some(result)) = self.sending_udp_v4_responses.poll_next_unpin(cx) {
+                result
+                    .context("Failed to send UDPv6 DNS response")
+                    .map_err(anyhow_to_io)?;
+
+                continue;
+            }
+
+            if let Poll::Ready(Some(result)) = self.reading_udp_v4_queries.poll_next_unpin(cx) {
+                let (from, message) = result
+                    .context("Failed to read UDPv4 DNS query")
+                    .map_err(anyhow_to_io)?;
+
+                return Poll::Ready(Ok(Query {
+                    source: from,
+                    message,
+                }));
+            }
+
+            if let Poll::Ready(Some(result)) = self.reading_udp_v6_queries.poll_next_unpin(cx) {
+                let (from, message) = result
+                    .context("Failed to read UDPv6 DNS query")
+                    .map_err(anyhow_to_io)?;
+
+                return Poll::Ready(Ok(Query {
+                    source: from,
+                    message,
+                }));
+            }
+
+            self.waker.register(cx.waker());
+            return Poll::Pending;
+        }
+    }
+}
+
+/// Produces a stream of incoming DNS queries from a UDP socket for as long as there is at least one strong reference to the socket.
+fn udp_dns_query_stream(
+    udp_socket: Weak<UdpSocket>,
+) -> BoxStream<'static, Result<(SocketAddr, Message<Vec<u8>>)>> {
+    stream::repeat(udp_socket) // We start with an infinite stream of weak references to the UDP socket.
+        .filter_map(|udp_socket| async move { udp_socket.upgrade() }) // For each item pulled from the stream, we first try to upgrade to a strong reference.
+        .then(read_udp_query) // And then read single DNS query from the socket.
+        .boxed()
+}
+
+fn anyhow_to_io(e: anyhow::Error) -> io::Error {
+    io::Error::other(format!("{e:#}"))
+}
+
+async fn read_udp_query(socket: Arc<UdpSocket>) -> Result<(SocketAddr, Message<Vec<u8>>)> {
+    let mut buffer = vec![0u8; 2000];
+
+    let (len, from) = socket
+        .recv_from(&mut buffer)
+        .await
+        .context("Failed to receive UDP packet")?;
+
+    buffer.truncate(len);
+
+    let message =
+        Message::try_from_octets(buffer).map_err(|_| anyhow!("Failed to parse DNS message"))?;
+
+    Ok((from, message))
```

Comment:
Would it be a good idea to calculate our max UDP packet size for this and enforce it strictly? Just thinking in case an edge case pops up where the length is longer than we expect.

Reply:
All DNS queries that arrive here should never be bigger than our MTU because they have been received on the TUN device.

Reply:
This crate doesn't know this however and we will reuse this for #8263 on the client. I added a comment that explains the choice of 2000.